docs/source/blueprints/BP003.rst

```diff
@@ -0,0 +1,366 @@
+BP003
+#####
+
+:Number: BP003
+:Title: N(ext)Runner Task Scheduler
+:Author: Cleber Rosa <crosa@redhat.com>
+:Discussions-To: avocado-devel@redhat.com
+:Reviewers:
+:Created: 20-July-2020
+:Type: Architecture Blueprint
+:Status: Draft
+
+.. contents:: Table of Contents
+
+TL;DR
+*****
+
+The N(ext) Runner has been used as Avocado's runner for selftests for
+over a year, but that has been implemented as ``avocado nrun``, that
+is, outside of the Avocado's traditional ``avocado run`` entrypoint.
+Under the hood, it means that the N(ext) Runner is not well integrated
+enough to an Avocado Job.
+
+A partial implementation of the N(ext) Runner integration to an
+Avocado Job is available at ``avocado/plugins/runner_nrunner.py`` but
+it has many limitations.  This blueprint describes the architecture of
+a scheduler that will power this Job compatible N(ext) Runner
+implementation.
+
+Motivation
+**********
+
+Describe the architecture of an task scheduler that will power the Job
+compatible N(ext) Runner implementation, allowing the N(ext) Runner
+first supported version to achieve feature completeness, and replace
+the current runner soon after the next LTS version is released.
```

Comment:
The answer to this "why" question has to do with what a decoupled, possibly distributed and parallel-capable runner needs to be/do.  The things the runner needs to be/do are so many, that I did not think it would fit in this section, for instance:

```
From a high-level view, the nrunner plugin needs to:

1. Break apart from the "one at a time" Task execution model that it currently
   employs;

2. Check if a Tasks can be executed, that is, if its requirements can be
   fulfilled (the most basic requirement for a task is a matching runner);

3. Prepare for the execution of a task, such as the fulfillment of extra task
   requirements. The requirements resolver is one, if not the only way,
   component that should be given a chance to act here;

4. Executes a task in a prepared environment;

5. Monitor the execution of a task (from an external PoV);

6. Collect the status messages that tasks will send;

  a. Forward the status messages to the appropriate job components, such as the
  result plugins.

  b. Depending on the content of messages, such as the ones containing "status:
  started" or "status: finished", interfere in the Task execution status, and
  consequently, in the Job execution status.

7. Verify, warn the user, and attempt to clean up stray tasks. This may be, for
   instance, necessary if a Task on a container seems to be stuck, and the
   container can not be destroyed. The same applies to the process in some time
   of uninterruptible periods of sleep.
```

So because of this large number of correlated responsibilities, there's the motivation of an entity such as an scheduler.  How do you think we can add the same info without duplicating too much?

Reply:
Yes, I do. IMO, the motivation should not be "describe the Scheduler architecture" but "why" we are proposing one. If the items on the section "requirements" are correlated to this motivation, so we just need to make some references here.

Reply:
I thing I get your point. I'll try to address it.